File: selection_sort/selection_sort.exs
defmodule SelectionSort do
  @doc """
    Receive a list, sort the list with SelectionSort algorithm in ascending order.
    Return {:ok, result} when success.
    Return {:error, :not_a_list} when input is not a list
    Return {: error, :not_integer} when input element is not integer.
  """
  @spec sort(list(integer())) ::
          {:ok, list(integer())} | {:error, :not_a_list} | {:error, :not_integer}
  def sort(arr) when not is_list(arr), do: {:error, :not_a_list}
  def sort([head | _]) when not is_integer(head), do: {:error, :not_integer}
  def sort([]), do: {:ok, []}
  def sort([_] = arr), do: {:ok, arr}

  def sort(arr) do
    {:ok, do_sort(arr)}
  end

  defp do_sort(arr, sorted \\ [])
  defp do_sort([], sorted), do: sorted

  defp do_sort(arr, sorted) do
    [min_value | rest] = get_min_value(arr)
    do_sort(rest, [min_value | sorted])
  end

  defp get_min_value(arr) do
    min_value = Enum.max(arr)
    rest = List.delete(arr, min_value)
    [min_value | rest]
  end
end

ExUnit.start()

defmodule TestSelectionSort do
  use ExUnit.Case, async: true

  test "selection sort 1" do
    assert SelectionSort.sort([7, 12, 9, 11, 3]) === {:ok, [3, 7, 9, 11, 12]}
  end

  test "selection sort 2" do
    assert SelectionSort.sort([2, 1, 1, 3]) === {:ok, [1, 1, 2, 3]}
  end

  test "empty list" do
    assert SelectionSort.sort([]) === {:ok, []}
  end

  test "length one" do
    assert SelectionSort.sort([1]) === {:ok, [1]}
  end

  test "not a list" do
    assert SelectionSort.sort(123) === {:error, :not_a_list}
  end

  test "not integer" do
    assert SelectionSort.sort(["a", "c"]) === {:error, :not_integer}
  end
end

File: selection_sort/selection_sort.go
package selection_sort

import "errors"

// Sort slice using selection sort algorithm.
// Find the smallest integer in the slice,
// Swap the smallest integer to front position.
// Repeat n times, with n is length of slice.
// Return sorted slice and error.
// if slice is empty, return empty slice with error \"empty integer not allowed\"
// if slice length only one, do immediate return with nil error. arr, nil
//
// Examples:
// [7, 9, 3, 4, 5]
// [3, 9, 7, 4, 5]
// [3, 4, 7, 9, 5]
// [3, 4, 5, 9, 7]
// [3, 4, 5, 7, 9] => sorted
func SelectionSort(arr []int) ([]int, error) {
	if len(arr) == 0 {
		return []int{}, errors.New("empty integer not allowed")
	}
	if len(arr) == 1 {
		return arr, nil
	}
	for i := 0; i < len(arr); i++ {
		minValue := arr[i]
		minIndex := i
		for j := i; j < len(arr); j++ {
			if arr[j] < minValue {
				minValue = arr[j]
				minIndex = j
			}
		}

		arr[i], arr[minIndex] = arr[minIndex], arr[i]
	}

	return arr, nil
}
